src/python/pants/backend/python/util_rules/faas.py

```diff
@@ -393,9 +442,106 @@ async def build_lambdex(
     return BuiltPackage(digest=result.output_digest, artifacts=(artifact,))
 
 
+@dataclass(frozen=True)
+class BuildPythonFaaSRequest:
+    address: Address
+    target_name: str
+
+    complete_platforms: PythonFaaSCompletePlatforms
+    handler: PythonFaaSHandlerField
+    output_path: OutputPathField
+    runtime: PythonFaaSRuntimeField
+    layout: Literal[PythonFaaSLayout.ZIP]
+
+    include_requirements: bool
+
+    reexported_handler_module: str
+    log_only_reexported_handler_func: bool = False
+
+
+@rule
+async def build_python_faas(
+    request: BuildPythonFaaSRequest,
+) -> BuiltPackage:
+    platform_str = request.runtime.to_platform_string()
+    pex_platforms = PexPlatforms([platform_str] if platform_str else [])
+
+    additional_pex_args = (
+        # Ensure we can resolve manylinux wheels in addition to any AMI-specific wheels.
+        "--manylinux=manylinux2014",
+        # When we're executing Pex on Linux, allow a local interpreter to be resolved if
+        # available and matching the AMI platform.
+        "--resolve-local-platforms",
+    )
+
+    complete_platforms, handler = await MultiGet(
+        Get(CompletePlatforms, PythonFaaSCompletePlatforms, request.complete_platforms),
+        Get(ResolvedPythonFaaSHandler, ResolvePythonFaaSHandlerRequest(request.handler)),
+    )
+
+    # TODO: improve diagnostics if there's more than one platform/complete_platform
+
+    # synthesise a source file that gives a fixed handler path, no matter what the entry point is:
+    # some platforms require a certain name (e.g. GCF), and even on others, giving a fixed name
+    # means users don't need to duplicate the entry_point config in both the pants BUILD file and
+    # infrastructure definitions (the latter can always use the same names, for every lambda).
+    reexported_handler_file = f"{request.reexported_handler_module}.py"
+    reexported_handler_func = "handler"
+    reexported_handler_content = (
+        f"from {handler.module} import {handler.func} as {reexported_handler_func}"
+    )
+    additional_sources = await Get(
+        Digest,
+        CreateDigest([FileContent(reexported_handler_file, reexported_handler_content.encode())]),
+    )
+
+    repository_filename = "faas_repository.pex"
+    pex_request = PexFromTargetsRequest(
+        addresses=[request.address],
+        internal_only=False,
```

Comment:
This feels unnecessarily heavyweight, and having to materialise the full non-internal PEX just to throw it away is presumably a major reason why this new approach is slower (since I imagine there's duplicated work between "build a PEX" and "turn that PEX into a venv"), related to the discussion in the PR.

However, this needs to be built for a potentially-foreign platform, and hence, AIUI, `internal_only=True` and `VenvPex` don't work.